Add clitests for the source command

mattmc3 opened the pull request:
- Add `.test` functions to aid test creation
- Fix #62 - test source pattern

Before submitting your PR (pull request), please
check the following:
* [x] There is no other PR (open or closed) similar to yours. If there is, please first discuss over there.
* [x] Each commit messages follows the [Seven Rules of a Great Commit Message](https://cbea.ms/git-commit/#seven-rules).
* [x] Each commit message includes `Fixes #<bug>` or `Resolves #<issue>` in its  body (not subject) for each issue it resolves (if any).
* [x] You have [squashed](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing) any redundant or insignificant commits.

Hey @marlonrichert!

I finally got around to converting my tests for [antidote](https://github.com/mattmc3/antidote) over to use [clitest](https://github.com/aureliojargas/clitest). We talked about this way back in 2021 on #22. Now that I know a lot more about `clitest`, I'm able to contribute to show how you might address #62. This PR adds some starter tests for your `znap source` command, as well as some test helper functions to help you expand testing going forward.

Your Zsh code style is completely different than mine, so I'm not sure if you'll even like what I have here (I tried to match you where I could). Regardless, I thought I'd pitch it to you anyway. No worries if you reject this PR and go your own way. I just figured I got a lot of milage out of `clitest` (great find!), and a lot of freedom to change things without worrying that stuff will break down the line due to excellent test coverage, so I thought you might benefit too.

I think testing like this also really becomes helpful as documentation for your users as to how Znap is intended to work, so I've tried to keep the tests relatively clean and put noise like `git` mocking and `mktemp` creation in the `.test.setup` func, and cleanup in the `.test.teardown` func so that the tests themselves are much more straight forward.

marlonrichert commented on .clitest/__init__.zsh:
```diff
@@ -2,4 +2,5 @@ zmodload zsh/param/private
 . scripts/opts.zsh
 setopt ${_znap_opts:#warn*}
 builtin autoload -Uz $PWD/functions/{znap,(|.).znap.*~*.zwc}
+builtin autoload -Uz $PWD/.clitest/functions/.test.*~*.zwc
```
Please don't autoload the test functions in production. 🙂 Move this code to somewhere where they get loaded only when testing.

mattmc3 replied:
Can you clarify? The `./.clitest/__init__.zsh` file is global, and not just for initializing tests? That seems like it's in the wrong folder then.

marlonrichert replied:
Ah, sorry, I was confusing it with [the other `__init__`](https://github.com/marlonrichert/zsh-autocomplete/blob/main/scripts/.autocomplete.__init__).

marlonrichert commented on .clitest/znap.source.md:
````diff
@@ -0,0 +1,40 @@
+Sourcing a path loads its plugin file:
+```zsh
+% mkdir -p ~znap/foo
+% echo "echo sourcing foo..." > ~znap/foo/foo.plugin.zsh
````
Why are you not using `.test.makeplugin` here?

mattmc3 replied:
Good catch.

marlonrichert commented on .clitest/znap.source.find.md:
```diff
@@ -0,0 +1,63 @@
+% .test.reset
+% dir=$T_TEMPDIR
+% mkdir -p $dir/foo
+% reply=(); .znap.source.find $dir/foo  #=> --exit 66
```
I didn't know you can check the exit status with `clitest`. Nice! 🙂

marlonrichert commented on .clitest/functions/.test.makeplugin:
```diff
@@ -0,0 +1,6 @@
+local file=${1:-foo.plugin.zsh}
+local plugin=${2:-${1%%.*}}
+mkdir -p $T_TEMPDIR/$plugin
+echo "echo 'sourcing ${plugin}/${file}...'" >$T_TEMPDIR/$plugin/$file
```
`print "print 'sourcing $plugin/$file...'"` would be more canonical in Zsh. Also, I prefer to have a space between `>` and a filename.

marlonrichert replied:
I would appreciate if you could still replace all instances of `echo` with `print -r`. 🙂

mattmc3 replied:
Sorry, missed that one.

marlonrichert commented on .clitest/functions/.test.setup:
```diff
@@ -0,0 +1,17 @@
+
+# works with BSD and GNU gmktemp
+T_TEMPDIR="$(mktemp -d -t t_znap.XXXXXXXX)"
+T_TEMPDIR=${T_TEMPDIR:A}
```
I would merge this with the line above and add spaces inside the brackets:
```zsh
T_TEMPDIR=${"$( mktemp -d -t t_znap.XXXXXXXX )":A}
```

marlonrichert commented on .clitest/functions/.test.setup:
```diff
@@ -0,0 +1,17 @@
+hash -d znap=$T_TEMPDIR
+
+# save fpath
+fpath_=( "$fpath[@]" )
```
The quotes and `[@]` are not necessary, unless you want to preserve empty strings:
```zsh
fpath_=( $fpath )
```

mattmc3 replied:
LOL, I copied that from your tests, so I figured that was your preferred style. Happy to change it, but do you want me to change it everywhere?

https://github.com/marlonrichert/zsh-snap/blob/9b8edd89c2b2303ffa253c5116a97d9138e0831d/.clitest/znap.prompt.load.md?plain=1#L4

marlonrichert commented on .clitest/functions/.test.setup:
```diff
@@ -0,0 +1,17 @@
+fpath_=( "$fpath[@]" )
+
+# mockgit
+git() { if [[ $1 == --version ]]; then print 2.8.0; else print $@[-1]; fi }
```
Please add line breaks. 🙂

marlonrichert commented on .clitest/functions/.test.teardown:
```diff
@@ -0,0 +1,10 @@
+#!/bin/zsh
+
+# restore git
+(( $+functions[git] )) && unfunction git
```
When using `&&` as an `if`, I prefer to add a line break after the `&&` plus indent the next line with 4 spaces. This makes your intention clearer and makes the diff easier to follow for Git, if you later decide to turn it into an actual `if`.

marlonrichert commented on .clitest/znap.source.find.md:
````diff
@@ -0,0 +1,63 @@
+```zsh
+% dir=$T_TEMPDIR
+% .test.makeplugin foo.plugin.zsh
+% reply=(); .znap.source.find $dir/foo  #=> --exit 0
````
I would prefer if you could use line breaks instead of `;` (except in constructs like `if ...; then`, `for ...; do`, etc.)

marlonrichert commented on .clitest/znap.source.find.md:
```diff
@@ -0,0 +1,63 @@
+% dir=$T_TEMPDIR
+% .test.makeplugin foo.plugin.zsh
+% reply=(); .znap.source.find $dir/foo  #=> --exit 0
+% echo ${reply[-1]:t}
```
Use `print -r -- $reply[-1]:t`. It's safer (ignores escape codes and doesn't interpret `-` as starting a flag) and more concise.

marlonrichert reviewed (changes requested):
Great work! 👏🏽 

Would you mind making the changes I request here? 🙂 I realize there's quite a few of them, but they are mostly simple find-replace actions.

mattmc3 commented:
> Great work! 👏🏽
> Would you mind making the changes I request here? 🙂 I realize there's quite a few of them, but they are mostly simple find-replace actions.

Happy to! I also tried not to touch your existing tests since I wasn't sure how you'd feel about this PR at first, so if you want me to clean those up too and make them all work the same, I'd be happy to do that. This structure should then give you something to build off of that will be sustainable as a testing suite across the whole project. Happy to help add more if you'd welcome the contribution.

marlonrichert commented on .clitest/functions/.test.setup:
```diff
@@ -0,0 +1,22 @@
+  if [[ $1 == --version ]]; then
+    print 2.8.0
+  else
+    print $@[-1]
```
Can you still make this `print -r -- $[@]`? Otherwise, an argument from a future test could break this.

marlonrichert commented on .clitest/functions/.test.teardown:
```diff
@@ -0,0 +1,12 @@
+#!/bin/zsh
+
+# restore git
+(( $+functions[git] )) &&
```
A better way to check this is `[[ -v functions[git] ]]`

marlonrichert commented on .clitest/functions/.test.teardown:
```diff
@@ -0,0 +1,12 @@
+
+# remove tempdir
+[[ -d "$T_TEMPDIR" ]] &&
+    rm -rf -- "$T_TEMPDIR"
```
The double quotes here are unnecessary. Zsh does not split on whitespace unless told to do so.

marlonrichert commented on functions/.znap.source.find:
```diff
@@ -9,6 +9,7 @@ if [[ -d $reply[1] ]]; then
 elif ! [[ -r $reply[1] ]]; then
   reply=( ${reply[1]}${^~ext}(NY1-.) )
 fi
+reply=(${(u)reply})
```
I tried to find where `.znap.source.find` is used and I couldn’t find it anywhere. I guess I forgot to remove this code at some point. 😅

Can you add a separate commit that removes it? And you can of course remove the test for it from your earlier commit.

mattmc3 replied:
That's funny! I couldn't find where it was being used either, but I assumed there must be a dynamic call I missed, and it had the most gnarly glob pattern so I assumed that was what #62 had to be referencing. I'll move the test concepts from `.znap.source.find` into `.znap.source` and we should be good to go.

mattmc3 commented:
I think I've addressed all the feedback. Let me know if I missed anything.

Also, it may be worth mentioning I do a couple other useful things in antidote's test suite that you might eventually find useful too in the future.

First, I save off/clear out/restore zstyles during setup and teardown. This prevents you from having to litter your tests with extra zstyle handling if you need to set a zstyle for a test and ensure it's cleared out after.

```zsh
### .test.setup
# save any existing zstyles, and clear them all for the test session
typeset -ga T_PREV_ZSTYLES=( ${(@f)"$(zstyle -L ':antidote:*')"} )
source <(zstyle -L ':znap:*' | awk '{print "zstyle -d", $2}')

### .test.teardown
# clear test zstyles and restore originals
source <(zstyle -L ':znap:*' | awk '{print "zstyle -d", $2}')
source <(printf '%s\n' $T_PREV_ZSTYLES)
```

Next, my clitests often test things that have non-deterministic output. Mostly that happens because of temp dirs. I think you will avoid some of those issues since you use `hash -d` directory shortcuts. But just in case, I also have found it handy to have a [`subenv` function](https://github.com/mattmc3/antidote/blob/main/tests/functions/subenv) that replaces non-deterministic output with variables.

```zsh
% T_TEMPDIR=$(mktemp -d)
% mydir=$T_TEMPDIR/foo/bar
% echo $mydir | subenv T_TEMPDIR
$T_TEMPDIR/foo/bar
```

`clitest` also supports regex matching of output, which might help non-deterministic output as well, but I've found it useful only for trivial cases.

```zsh
$ echo $((2 + 10))  #=> --regex ^\d+$
```

Obviously, if you figure out better ways to handle these things or any other tricks with `clitest`, I'm happy to learn from your experience as well.

mattmc3 commented:
@marlonrichert - Is this a PR you decided not to merge after all? I'm looking to clean up some of my projects and wanted delete this fork, but don't want you to lose this if it's stuff you want.

marlonrichert commented:
@mattmc3 Sorry, I completely forgot about this. I've been too busy with other things.

Looks good to me. I'll merge it.
